enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,1169 @@
+    ComputeClusterFailureDomain VSpherePlatformFailureDomainType = "ComputeCluster"
+    DatacenterFailureDomain     VSpherePlatformFailureDomainType = "Datacenter"
+)
+
+// VSpherePlatformFailureDomainSpec holds the region and zone failure domain and
+// the vCenter topology of that failure domain.
+type VSpherePlatformFailureDomainSpec struct {
+  // name defines the name of the FailureDomainSpec
+  // This name is abritary but will be used
+  // in VSpherePlatformDeploymentZone for association.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  Name string `json:"name"`
+
+  // region defines the VSpherePlatformFailureDomain
+  // which contains
+  // +kubebuilder:validation:Required
+  Region VSpherePlatformFailureDomain `json:"region"`
+
+  // Zone defines the name and type of a zone
+  // +kubebuilder:validation:Required
+  Zone VSpherePlatformFailureDomain `json:"zone"`
+
+  // Topology describes a given failure domain using vSphere constructs
+  // +kubebuilder:validation:Required
+  Topology VSpherePlatformTopology `json:"topology"`
+}
+
+// VSpherePlatformFailureDomain holds the name of the associated tag, the type
+// of the failure domain, and the vCenter tag category associated with this
+// failure domain.
+type VSpherePlatformFailureDomain struct {
+  // name is the name of the vCenter tag that represents this failure domain
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  Name string `json:"name"`
+
+  // type is the type of failure domain, the current values are "Datacenter" and "ComputeCluster"
+  // +kubebuilder:validation:Enum=HostGroup;Datacenter;ComputeCluster
+  // +kubebuilder:validation:Required
+  Type VSpherePlatformFailureDomainType `json:"type"`
+
+  // tagCategory is the category used for the tag
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  TagCategory string `json:"tagCategory"`
+}
+
+// VSpherePlatformTopology holds the required and optional vCenter objects - datacenter,
+// computeCluster, networks, datastore and resourcePool - to provision virtual machines.
+type VSpherePlatformTopology struct {
+  // datacenter is the vCenter datacenter in which virtual machines will be located
+  // and defined as the failure domain.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  Datacenter string `json:"datacenter"`
+
+  // computeCluster as the failure domain
+  // TODO: verify maximum length
+  // TODO: what if this is a inventory path?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  ComputeCluster *string `json:"computeCluster,omitempty"`
+
+  // Hosts has information required for placement of machines on VSphere hosts.
```

Comment:
the use of `required` makes `optional` nature of this field seem ambiguous to me